### ISSUE TITLE
aws: guardduty datastream collects duplicated documents

### ISSUE THREAD

**Issue by efd6** [elastic/integrations#8601]
We have a case where the GuardDuty datastream is collecting documents more than once. This appears to be due to the API returning the same pagination cursor more than once in a pagination sequence.

For example in the pagination sequence that starts from this seed request (identifying details obfuscated in all quoted logs):
```
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:12:21.233Z",
  "message": "HTTP request",
  "transaction.id": "... -391",
  "url.original": "https://guardduty. ... .amazonaws.com/detector/.../findings",
  "url.scheme": "https",
  "url.path": "/detector/.../findings",
  "url.domain": "guardduty. ... .amazonaws.com",
  "url.port": "",
  "url.query": "",
  "http.request.method": "POST",
  "user_agent.original": "...",
  "http.request.body.content": "{\"findingCriteria\":{\"criterion\":{\"updatedAt\":{\"greaterThan\":\"1701140400000\",\"lessThan\":\"1701144000000\"}}},\"maxResults\":50,\"sortCriteria\":{\"attributeName\":\"updatedAt\",\"orderBy\":\"ASC\"}}",
  "http.request.body.bytes": 183,
  "http.request.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
```
We get the following request/response/request sequence (note response's `http.response.body.content` and `http.response.body.bytes`, and the `http.request.body.content` in the final request):
```
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:12:22.216Z",
  "message": "HTTP request",
  "transaction.id": "... -395",
  "url.original": "https://guardduty. ... .amazonaws.com/detector/.../findings",
  "url.scheme": "https",
  "url.path": "/detector/.../findings",
  "url.domain": "guardduty. ... .amazonaws.com",
  "url.port": "",
  "url.query": "",
  "http.request.method": "POST",
  "user_agent.original": "...",
  "http.request.body.content": "{\"findingCriteria\":{\"criterion\":{\"updatedAt\":{\"greaterThan\":\"1701140400000\",\"lessThan\":\"1701144000000\"}}},\"maxResults\":50,\"nextToken\":\"1701142787842-1701142787000-md5sum(4b32102b95b3986057cc16a660419069)\",\"sortCriteria\":{\"attributeName\":\"updatedAt\",\"orderBy\":\"ASC\"}}",
  "http.request.body.bytes": 258,
  "http.request.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:12:22.316Z",
  "message": "HTTP response",
  "transaction.id": "... -395",
  "http.response.status_code": 200,
  "http.response.body.content": md5sum(1d166755ca9ad65949bf55eee7e9b181),
  "http.response.body.bytes": 1841,
  "http.response.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:12:22.518Z",
  "message": "HTTP request",
  "transaction.id": "... -396",
  "url.original": "https://guardduty. ... .amazonaws.com/detector/.../findings",
  "url.scheme": "https",
  "url.path": "/detector/.../findings",
  "url.domain": "guardduty. ... .amazonaws.com",
  "url.port": "",
  "url.query": "",
  "http.request.method": "POST",
  "user_agent.original": "...",
  "http.request.body.content": "{\"findingCriteria\":{\"criterion\":{\"updatedAt\":{\"greaterThan\":\"1701140400000\",\"lessThan\":\"1701144000000\"}}},\"maxResults\":50,\"nextToken\":\"1701142787857-1701142787000-md5sum(fcc045188185e1c27e14fdc6b66e2493)\",\"sortCriteria\":{\"attributeName\":\"updatedAt\",\"orderBy\":\"ASC\"}}",
  "http.request.body.bytes": 258,
  "http.request.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
```
In the same pagination sequence (i.e. not after a new seed request), we get the following set of three logs that show exactly the same `http.response.body.content` and `http.request.body.content`.
```
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:27:22.110Z",
  "message": "HTTP request",
  "transaction.id": "... -401",
  "url.original": "https://guardduty. ... .amazonaws.com/detector/.../findings",
  "url.scheme": "https",
  "url.path": "/detector/.../findings",
  "url.domain": "guardduty. ... .amazonaws.com",
  "url.port": "",
  "url.query": "",
  "http.request.method": "POST",
  "user_agent.original": "...",
  "http.request.body.content": "{\"findingCriteria\":{\"criterion\":{\"updatedAt\":{\"greaterThan\":\"1701140400000\",\"lessThan\":\"1701144000000\"}}},\"maxResults\":50,\"nextToken\":\"1701142787842-1701142787000-md5sum(4b32102b95b3986057cc16a660419069)\",\"sortCriteria\":{\"attributeName\":\"updatedAt\",\"orderBy\":\"ASC\"}}",
  "http.request.body.bytes": 258,
  "http.request.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:27:22.219Z",
  "message": "HTTP response",
  "transaction.id": "... -401",
  "http.response.status_code": 200,
  "http.response.body.content": md5sum(1d166755ca9ad65949bf55eee7e9b181),
  "http.response.body.bytes": 1841,
  "http.response.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
{
  "log.level": "debug",
  "@timestamp": "2023-11-28T04:27:22.417Z",
  "message": "HTTP request",
  "transaction.id": "... -402",
  "url.original": "https://guardduty. ... .amazonaws.com/detector/.../findings",
  "url.scheme": "https",
  "url.path": "/detector/.../findings",
  "url.domain": "guardduty. ... .amazonaws.com",
  "url.port": "",
  "url.query": "",
  "http.request.method": "POST",
  "user_agent.original": "...",
  "http.request.body.content": "{\"findingCriteria\":{\"criterion\":{\"updatedAt\":{\"greaterThan\":\"1701140400000\",\"lessThan\":\"1701144000000\"}}},\"maxResults\":50,\"nextToken\":\"1701142787857-1701142787000-md5sum(fcc045188185e1c27e14fdc6b66e2493)\",\"sortCriteria\":{\"attributeName\":\"updatedAt\",\"orderBy\":\"ASC\"}}",
  "http.request.body.bytes": 258,
  "http.request.mime_type": "application/json",
  "ecs.version": "1.6.0"
}
```
In the case where this is observed, this result it duplicated documents in the index. This should not be happening due to fingerprinting, but that is independent of this. We should not be receiving/collecting duplicated documents. If this is a limitation of the API we need to find a work around and if we are holding the API incorrectly, we need to fix that.

**Comment by efd6** [comment]
Taking a fresh look at this and it appears that the two sequences are not part of the same cursor pagination sequence. This is indicated by the timestamps on the two log excerpts, being 15 minutes apart.

On the basis of this, it looks like this is an issue with last execution template evaluation.

The issue appears to be due to a one-hour truncation [here](https://github.com/elastic/integrations/blob/c91fc59e6ce1ac19f55c775ed5e1e52fa870c20c/packages/aws/data_stream/guardduty/agent/stream/httpjson.yml.hbs#L32). The offset in the expression may be less than an hour (15m in the case of the system that the original issue was identified) which would result in refetch of the same data for ⌈60/interval⌉ times where interval is minutes (under 60). It's unclear to me why there should be a truncation there.

**Comment by efd6** [comment]
In the PR I wrote "Not marking as closing the issue as I think there are still problems with the template evaluation that are not currently diagnosable with HTTPJSON's existing template logging (will be able to look at this when v8.11 is more prevalent )." Nothing has changed except that v8.11+ will be more prevalent and so if we see the issue it will be easier to debug. We could close this now if you want, with a view to either reopening or opening a new issue if there is a new report.

**Comment by efd6** [comment]
I have an answer about ["It's unclear to me why there should be a truncation there."](https://github.com/elastic/integrations/issues/8601#issuecomment-1891226889)

From Crest:

> The AWS API is restrictive, particularly in that we must keep the `greaterThan` and `lessThan` parameters exactly the same across all pagination calls; otherwise, the AWS API returns an error.
>
> In the beginning of data collection, we used the current time (`now`) here to set these parameters for the first call. However, by the time we made subsequent pagination requests, `now` had changed, resulting in errors because there was no way to retrieve the date from the last request's body. To resolve this issue, we've truncated the date to the nearest hour, which has corrected the problem.

This means that the truncation would not be required if the data stream were re-written to use CEL, since there we can keep the timestamp value across all pagination requests by including it in the returned state.

**Comment by ShourieG** [comment]
@efd6 Since GuardDuty is a highly popular and widely used integration, do you feel a CEL re-write is the right approach or should we introduce a 2nd CEL input as an option ? If neither, then we can stick to the current change made by Crest and close the issue. Otherwise we can create a new issue for the required change.

**Comment by efd6** [comment]
There is no current fix for this; that reference to Crest was just a clarification of why the integration template was written the way it was. It is not a fix for the issue that is noted here.

I don't think that we should have two versions of the integration in parallel since we have no clear deprecation/removal path for the HTTP JSON version. Where we _have_ done this it has been for cases where there are deprecated API versions (the Mimecast integration is the perfect example of this), but that's not the case here.

The balance of costs of migration versus ongoing misbehaviour is really a PM decision, but the issue should not be closed until the issue is fixed or otherwise deprioritised.

cc @jamiehynds

**Comment by jamiehynds** [comment]
> The balance of costs of migration versus ongoing misbehaviour is really a PM decision, but the issue should not be closed until the issue is fixed or otherwise deprioritised.


@efd6  - when you say misbehaviour, are we referring to the continued ingestion and storage of duplicate documents, with migration to CEL being the only current workaround? Or have we mitigated the issue to some degree, and CEL simply provides a more robust long-term solution?

Do we have any sense of how prevalent this issue is or how likely users are to encounter the duplication problem? Since we’ve only seen a single report so far (as far as I can tell), my inclination is that the migration cost to CEL, combined with the usual complexity of maintaining both httpjson and CEL inputs, doesn’t seem justified at this stage.

**Comment by efd6** [comment]
> Or have we mitigated the issue to some degree, and CEL simply provides a more robust long-term solution?

This (it's matter of degree).

> Do we have any sense of how prevalent this issue is or how likely users are to encounter the duplication problem?

Unfortunately not.

> Since we’ve only seen a single report so far (as far as I can tell), my inclination is that the migration cost to CEL, combined with the usual complexity of maintaining both httpjson and CEL inputs, doesn’t seem justified at this stage.

OK.

@ShourieG Maybe we can document the current limitation in the readme?

**Comment by ShourieG** [comment]
@efd6, could you elaborate a bit on the documentation update that needs to be made.

**Comment by efd6** [comment]
The documentation would be a warning that the problem exists and is known. We can not offer a way to mitigate since if we could we could just have fixed it. This is wholly unsatisfactory, but it's the option that we have.

**Comment by efd6** [comment]
For the sake of clarity I've added a checklist for the steps that we need(ed) to address this issue.
